Dockerfile - Cryptography build fails without libffi-dev package

Cryptography build fails without libffi-dev package.

Partial error output for `docker build -t jsnapy .`

```
   x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/src/lxml/lxml.objectify.o -lxslt -lexslt -lxml2 -lz -lm -o build/lib.linux-x86_64-2.7/lxml/objectify.so

  Running setup.py install for cryptography
    c/_cffi_backend.c:15:17: fatal error: ffi.h: No such file or directory
     #include <ffi.h>
                     ^
    compilation terminated.
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip_build_root/cryptography/setup.py", line 333, in <module>
        **keywords_with_side_effects(sys.argv)
      File "/usr/lib/python2.7/distutils/core.py", line 111, in setup
        _setup_distribution = dist = klass(attrs)
      File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line 239, in __init__
        self.fetch_build_eggs(attrs.pop('setup_requires'))
      File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line 264, in fetch_build_eggs
        replace_conflicting=True
      File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 620, in resolve
        dist = best[req.key] = env.best_match(req, ws, installer)
      File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 858, in best_match
        return self.obtain(req, installer) # try and download/install
      File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 870, in obtain
        return installer(requirement)
      File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line 314, in fetch_build_egg
        return cmd.easy_install(req)
      File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 616, in easy_install
        return self.install_item(spec, dist.location, tmpdir, deps)
      File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 646, in install_item
        dists = self.install_eggs(spec, download, tmpdir)
      File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 834, in install_eggs
        return self.build_and_install(setup_script, setup_base)
      File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 1040, in build_and_install
        self.run_setup(setup_script, setup_base, args)
      File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 1028, in run_setup
        raise DistutilsError("Setup script exited with %s" % (v.args[0],))
    distutils.errors.DistutilsError: Setup script exited with error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    Complete output from command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/cryptography/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-EzxxXN-record/install-record.txt --single-version-externally-managed --compile:
    c/_cffi_backend.c:15:17: fatal error: ffi.h: No such file or directory

 #include <ffi.h>

                 ^

compilation terminated.

Traceback (most recent call last):

  File "<string>", line 1, in <module>

  File "/tmp/pip_build_root/cryptography/setup.py", line 333, in <module>

    **keywords_with_side_effects(sys.argv)

  File "/usr/lib/python2.7/distutils/core.py", line 111, in setup

    _setup_distribution = dist = klass(attrs)

  File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line 239, in __init__

    self.fetch_build_eggs(attrs.pop('setup_requires'))

  File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line 264, in fetch_build_eggs

    replace_conflicting=True

  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 620, in resolve

    dist = best[req.key] = env.best_match(req, ws, installer)

  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 858, in best_match

    return self.obtain(req, installer) # try and download/install

  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 870, in obtain

    return installer(requirement)

  File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line 314, in fetch_build_egg

    return cmd.easy_install(req)

  File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 616, in easy_install

    return self.install_item(spec, dist.location, tmpdir, deps)

  File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 646, in install_item

    dists = self.install_eggs(spec, download, tmpdir)

  File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 834, in install_eggs

    return self.build_and_install(setup_script, setup_base)

  File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 1040, in build_and_install

    self.run_setup(setup_script, setup_base, args)

  File "/usr/lib/python2.7/dist-packages/setuptools/command/easy_install.py", line 1028, in run_setup

    raise DistutilsError("Setup script exited with %s" % (v.args[0],))

distutils.errors.DistutilsError: Setup script exited with error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

----------------------------------------
Cleaning up...
Command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/cryptography/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-EzxxXN-record/install-record.txt --single-version-externally-managed --compile failed with error code 1 in /tmp/pip_build_root/cryptography
Storing debug log for failure in /root/.pip/pip.log
The command '/bin/sh -c pip install ncclient' returned a non-zero code: 1
root@ubuntu-1gb-nyc1-01:~/jsnapy#
```

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,15 +1,6 @@
-FROM ubuntu:14.04
+FROM juniper/pyez
 
-RUN apt-get -y update && \
-    apt-get -y upgrade && \
-    apt-get install -y wget build-essential libwrap0-dev libssl-dev \
-         python-distutils-extra libc-ares-dev uuid-dev python-pip \
-         python-dev libxml2-dev libxslt1-dev libffi-dev ansible libyaml-dev
-
-#1 install Pyez & update distribute package
-RUN pip install -U ncclient junos-eznc distribute
-
-#2 install jsnapy
+#1 install jsnapy
 RUN mkdir -p /usr/local/tmp/jsnapy
 ADD lib /usr/local/tmp/jsnapy/lib
 ADD logs /usr/local/tmp/jsnapy/logs
@@ -24,15 +15,18 @@ WORKDIR /usr/local/tmp/jsnapy
 RUN python setup.py sdist
 RUN pip install dist/jsnapy-0.1.tar.gz
 
-#3 install  netconify
+#2 install  netconify
 RUN pip install junos-netconify
 
-#4 - Install Junos Galaxy on ansible
-RUN ansible-galaxy install Juniper.junos
+#3 - Install ansible and Junos Galaxy module
+RUN apk update \
+    && apk upgrade \
+    && apk add ansible \
+    && ansible-galaxy install Juniper.junos
 
-#5 - Clean up
-RUN apt-get clean &&\
-apt-get purge
+#4 - Clean up
+RUN rm -rf /var/cache/apk/*
 
-#6 Optional - Copy example files
+#5 Optional - Copy example files
 #ADD ./example /tmp/example
+

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,23 +1,25 @@
 FROM ubuntu:14.04
 
 RUN apt-get -y update && \
-    apt-get -y upgrade
+    apt-get -y upgrade && \
+    apt-get install -y wget build-essential libwrap0-dev libssl-dev \
+         python-distutils-extra libc-ares-dev uuid-dev python-pip \
+         python-dev libxml2-dev libxslt1-dev libffi-dev ansible libyaml-dev
 
-RUN apt-get install -y wget build-essential libwrap0-dev libssl-dev python-distutils-extra libc-ares-dev uuid-dev
-RUN apt-get install -y python-pip python-dev libxml2-dev libxslt1-dev
+#1 install Pyez & update distribute package
+RUN pip install -U ncclient junos-eznc distribute
 
-#1 install Pyez
-RUN pip install ncclient
-RUN pip install junos-eznc
-
-#2 install jsnapy 
+#2 install jsnapy
 RUN mkdir -p /usr/local/tmp/jsnapy
 ADD lib /usr/local/tmp/jsnapy/lib
-ADD setup.py /usr/local/tmp/jsnapy/setup.py
-ADD tests /usr/local/tmp/jsnapy/tests
+ADD logs /usr/local/tmp/jsnapy/logs
 ADD MANIFEST.in /usr/local/tmp/jsnapy/MANIFEST.in
 ADD requirements.txt /usr/local/tmp/jsnapy/requirements.txt
+ADD samples /usr/local/tmp/jsnapy/samples
+ADD setup.py /usr/local/tmp/jsnapy/setup.py
 ADD snapshots /usr/local/tmp/jsnapy/snapshots
+ADD testfiles /usr/local/tmp/jsnapy/testfiles
+ADD tools /usr/local/tmp/jsnapy/tools
 WORKDIR /usr/local/tmp/jsnapy
 RUN python setup.py sdist
 RUN pip install dist/jsnapy-0.1.tar.gz
@@ -25,16 +27,12 @@ RUN pip install dist/jsnapy-0.1.tar.gz
 #3 install  netconify
 RUN pip install junos-netconify
 
-#4 Install Ansible
-RUN mkdir -p /etc/ansible
-RUN pip install ansible 
-
-#5 - Install Junos Galaxy on ansible
+#4 - Install Junos Galaxy on ansible
 RUN ansible-galaxy install Juniper.junos
 
-#6 - Clean up
-RUN apt-get clean &&\ 
+#5 - Clean up
+RUN apt-get clean &&\
 apt-get purge
 
-#7 Optional - Copy example files
+#6 Optional - Copy example files
 #ADD ./example /tmp/example

